Implement `PhoneNumberCard` and `ContactInformation` components for contact info

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Partial of #863 and based on #914

- Add `libphonenumber-js` to dependencies
- Add hooks to get all supported country calling codes and user's MC phone number
- Add phone number appearance and 'hideIcon' prop to **AccountCard**, and make its className extendable
- Decode HTML entities within country names
- Implement the shared components of **PhoneNumberCard** and **ContactInformation** section

💡  Please note that this PR doesn't include the API integration.

### Screenshots:

#### 🎥 . When the phone number got from Google MC is empty

https://user-images.githubusercontent.com/17420811/127475223-8cfbcb18-4d78-4d6b-86f5-fb4f2f1ee3e2.mp4

#### 🎥 . When the phone number got from Google MC has already exist

https://user-images.githubusercontent.com/17420811/127475258-e8a300b1-3d40-401e-9867-671a4506b80d.mp4

#### 📷 . The preview mode of PhoneNumberCard will be used in the Setting page later

![image](https://user-images.githubusercontent.com/17420811/127476444-28217c95-a4a0-40a3-9cb7-678cfc810a89.png)

### Detailed test instructions:

For easier testing, there is a mock phone number that is 50% chance for an empty phone number and 50% for an existed phone number when every time reload the page.

https://github.com/woocommerce/google-listings-and-ads/blob/4df92ad07f4c6f156a33988fe7be2115169f7f6b/js/src/data/selectors.js#L71-L78

1. Go to the MC setup flow and proceed to the "Confirm store requirements" step
2. The phone number card should be able to interact
3. The country calling code and national number should be printed on DevTool console when changing it via UI 
    ![image](https://user-images.githubusercontent.com/17420811/127476932-d15eac24-04d1-483b-80b1-db7691bfaa2f.png)

### Changelog entry

eason9487 committed:
Add 'libphonenumber-js' to dependencies

eason9487 committed:
Decode HTML entities within countey names

eason9487 committed:
Add a hook to get all supported country calling codes

eason9487 committed:
Add a hook to get user's MC phone number

eason9487 committed:
Add 'hideIcon' prop to AccountCard and make its className extendable

eason9487 committed:
Add new appearance to AccountCard for phone number

eason9487 committed:
Implement the shared components of PhoneNumberCard and ContactInformation section

tomalec commented:
> Add `libphonenumber-js` to dependencies

Hopefully, it won't get taken down again or messed around due to some controversies, https://medium.com/@catamphetamine/how-github-blocked-me-and-all-my-libraries-c32c61f061d3.


> For easier testing, there is a mock phone number that is 50% chance for an empty phone number and 50% for an existed phone number when every time reload the page.

:heart_eyes: Thanks

tomalec commented on js/src/components/contact-information/phone-number-card.js:
```diff
@@ -0,0 +1,144 @@
+	const [ country, setCountry ] = useState( initCountry );
+	const [ number, setNumber ] = useState( initNationalNumber );
+
+	const handleChange = ( nextCountry, nextNumber ) => {
+		if ( nextCountry !== country ) {
+			setCountry( nextCountry );
+		}
+		if ( nextNumber !== number ) {
+			setNumber( nextNumber );
+		}
+
+		const countryCallingCode = nextCountry
+			? getCountryCallingCode( nextCountry )
+			: '';
+		onPhoneNumberChange( countryCallingCode, nextNumber, nextCountry );
+	};
+
+	const handleCountryChange = ( v ) => handleChange( v, number );
+	const handleNumberChange = ( v ) => handleChange( country, v );
```
:nail_care: 
- Personally, I try to avoid single-char variable names, as they are less descriptive than full names, harder to `ctrl + f`, sometimes ambiguous.
- Conditions at l.35 and l.38 are know in advance and are de-facto differentiating `handleCountryChange` vs. `handleNumberChange` again. Which is already done.
  WDYT of something like:
  ```js	
	const handlePhoneChange = ( nextCountry, nextNumber ) => {
		const countryCallingCode = nextCountry
			? getCountryCallingCode( nextCountry )
			: '';
		onPhoneNumberChange( countryCallingCode, nextNumber, nextCountry );
	};
	const handleCountryChange = ( nextCountry ) => {
		if ( nextCountry !== country ) {
			setCountry( nextCountry );
		}

		handlePhoneChange( number, nextCountry );
	};
	const handleNumberChange = ( nextNumber ) => {
		if ( nextNumber !== number ) {
			setNumber( nextNumber );
		}

		handlePhoneChange( nextNumber, country );
	};
	```
- [`setState` docs](https://reactjs.org/docs/hooks-reference.html#usestate) state
	> If your update function returns the exact same value as the current state, the subsequent rerender will be skipped completely.
	
	So, if I understand that correctly, `if ( nextSate !== currentState ) {  setState( nextState ); }` is already covered.
	Then we could make it 
	```js
 	const handlePhoneChange = ( nextCountry, nextNumber ) => {
		setCountry( nextCountry );
		setNumber( nextNumber );

		const countryCallingCode = nextCountry
			? getCountryCallingCode( nextCountry )
			: '';
		onPhoneNumberChange( countryCallingCode, nextNumber, nextCountry );
	};

	const handleCountryChange = ( nextCountry ) =>
		handlePhoneChange( nextCountry, number );
	const handleNumberChange = ( nextNumber ) =>
		handlePhoneChange( country, nextNumber );
	```

eason9487 replied:
Changed by https://github.com/woocommerce/google-listings-and-ads/pull/917/commits/8f42c78e10794eed924ff97b0c1a671552231880

tomalec commented on js/src/hooks/useCountryKeyNameMap.js:
```diff
@@ -4,6 +4,18 @@
 import { SETTINGS_STORE_NAME } from '@woocommerce/data';
 import { useSelect } from '@wordpress/data';
 
+/*
+ * Examples:
+ * - Cura&ccedil;ao                             -> Curaçao
+ * - Saint Barth&eacute;lemy                    -> Saint Barthélemy
+ * - S&atilde;o Tom&eacute; and Pr&iacute;ncipe -> São Tomé and Príncipe
+ */
+function replaceHtmlEntities( countryName ) {
+	const el = document.createElement( 'span' );
+	el.innerHTML = countryName;
+	return el.textContent;
+}
+
```
:nail_care: WDYT of using https://www.npmjs.com/package/@wordpress/html-entitiesm? It's somewhat test-covered.

eason9487 replied:
Replaced by https://github.com/woocommerce/google-listings-and-ads/pull/917/commits/2a59a994a56802d1b828c05136cab28683757ced

tomalec commented on js/src/hooks/useCountryKeyNameMap.js:
```diff
@@ -18,10 +30,19 @@ import { useSelect } from '@wordpress/data';
+
+		Object.keys( countries ).forEach( ( key ) => {
+			const name = countries[ key ];
+			if ( /&[^;]+;/.test( name ) ) {
```
[`@wordpress/html-entities`](https://www.npmjs.com/package/@wordpress/html-entitieshttps://www.npmjs.com/package/@wordpress/html-entities) has a check like this built in: https://github.com/WordPress/gutenberg/blob/f53785fbbf674d46101831206a5154d095e39d0d/packages/html-entities/src/index.js#L19

eason9487 replied:
Replaced by https://github.com/woocommerce/google-listings-and-ads/pull/917/commits/2a59a994a56802d1b828c05136cab28683757ced

tomalec commented on js/src/hooks/useCountryCallingCodes.js:
```diff
@@ -0,0 +1,38 @@
+	label: `${ countryName } (+${ countryCallingCode })`,
+} );
+
+export default function useCountryCallingCodes( as ) {
```
:question: / :scroll:
Do you envision some more formats, as it's a string? or could we make it `asOptions = false`, to make the available options/values more explicit.
Also, I have't found any usage of `useCountryCallingCodes( 'to-data' )` or `useCountryCallingCodes( anythingelse )`.
Do we even need that parameter?

eason9487 replied:
I created the hook with the general name `useCountryCallingCodes` and thought its data structure looks too particular if returns option structure directly. So I added the `as` parameter to indicate the returning data structure.

But considering it will probably only be used with this data structure, I renamed and simplified it by https://github.com/woocommerce/google-listings-and-ads/pull/917/commits/6e839d8b3b325fd85c5952d9bb0c1f48dc98887b

tomalec commented:
Speaking of [potential problems](https://github.com/woocommerce/google-listings-and-ads/pull/917#issuecomment-891031409) with dependencies, do we have any policy of evaluating the dependencies we use? Especially, that here, we feed out customers' phone numbers to external software.
That could lead to some GDPR/PII issues.
//cc @jconroy 


BTW, :scroll: I don't think we need to address this here, or is it a strict blocking issue. We already process our code through an endless amount of don't-know-how-well-audited dependencies - Webpack & a handful of plugins.

tomalec commented on js/src/components/contact-information/phone-number-card.scss:
```diff
@@ -0,0 +1,42 @@
+			}
+
+			.woocommerce-select-control__option {
+				min-height: 36px;
+				font-size: 13px;
```
I was thinking of using some variables shared with `woocommerce/components` here, to keep it consistent and futureproof, but it seems `13` and `36` are hardcoded there as well :|

tomalec commented on js/src/components/contact-information/phone-number-card.js:
```diff
@@ -0,0 +1,144 @@
+	onPhoneNumberChange = noop,
+} ) {
+	const [ isEditing, setEditing ] = useState( initEditing );
+	const { loaded, data } = useGoogleMCPhoneNumber();
```
:nail_care: Having it here and in a containing element, https://github.com/woocommerce/google-listings-and-ads/pull/917/files#diff-51ff763be0b20871d193755311c5c437caafd6c4d46bb1413de599e1124a7714R25 seems redundant.

Why don't we make this a dumb component that gets fed with data and calls `*Change` callback?

eason9487 replied:
Changed by https://github.com/woocommerce/google-listings-and-ads/pull/917/commits/2a5ab61de8a0433ecea71218abc2fce914b3ce15

tomalec commented on js/src/components/contact-information/phone-number-card.js:
```diff
@@ -0,0 +1,144 @@
+			'google-listings-and-ads'
+		);
+
+		if ( loaded ) {
+			phoneNumberContent = (
+				<>
+					<CardDivider />
+					<PhoneNumberContent
+						initCountry={ data.country }
+						initNationalNumber={ data.nationalNumber }
+						onPhoneNumberChange={ onPhoneNumberChange }
+					/>
+				</>
+			);
+		} else {
+			phoneNumberContent = <AppSpinner />;
+		}
+	} else {
+		description = loaded ? data.display : <Spinner />;
```
:question:  Having the preloading logic here and in https://github.com/woocommerce/google-listings-and-ads/pull/917/files#diff-51ff763be0b20871d193755311c5c437caafd6c4d46bb1413de599e1124a7714R53 looks like too much.
I think we can make containing element render this component. Then, this component could decide whether to render a scaffold with a tiny preloader in place of a phone number, or a fully prepared input control.

Also, in one case we use `<Spinner>` in the other `<AppSpinner>`. Is that on purpose?

eason9487 replied:
Sorry, I'm not quite sure of your meaning. Could you provide some code examples?

The logic is used for rendering different visual results.

![Kapture 2021-08-03 at 11 36 33](https://user-images.githubusercontent.com/17420811/127955340-9827b893-bf42-4fef-b8d2-5cc29445223e.gif)

The `<Spinner>` is shown in the first, and the `<AppSpinner>` is for the second one.

tomalec replied:
Sorry, I put the wrong link.
My main point was about handling loaded here and in `ContactInformation`
https://github.com/woocommerce/google-listings-and-ads/pull/917/files#diff-51ff763be0b20871d193755311c5c437caafd6c4d46bb1413de599e1124a7714R53
what results in 
![image](https://user-images.githubusercontent.com/17435/127996658-bf259ca4-7721-421e-9fe5-11262de8ec5f.png)

Here is code sample/suggestion https://github.com/woocommerce/google-listings-and-ads/pull/917/files#r681624815

tomalec replied:
Speaking of `<Spinner>` vs. `<AppSpinner>` now I got it, thanks. I agree we need both here. I got mislead by their names. I thought `App-` one is just profiled for our need, while they are in fact `<InlineSpinenr>` and `<BlockSpinner>`.

tomalec replied:
:nail_care: :scroll: For the block spinner `<AppSpinner>`, I'd adapt its position/margin/padding to make the card fixed height = do not jump once loaded.

eason9487 replied:
I changed the loading UI in the later PR #921, and it shows a spinner with a card layout instead. The demo in another PR shows that change:

https://user-images.githubusercontent.com/17420811/127852973-dcbed8d8-f2ea-4dc5-91f4-85e66199617d.mp4

tomalec reviewed (commented):
Tested locally, reviewed the code.
Works fine, but left few comments regarding the code improvements.

eason9487 committed:
Rename parameter and simplify conditional `setState` calls

Address https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681013079

eason9487 committed:
Use @wordpress/html-entities instead

Address
- https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681036085
- https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681038414

eason9487 committed:
Rename useCountryCallingCodes to useCountryCallingCodeOptions and remove `as` parameter

Address https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681073767

eason9487 committed:
Remove useGoogleMCPhoneNumber from PhoneNumberCard

Address https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681154403

tomalec commented on js/src/components/contact-information/index.js:
```diff
@@ -0,0 +1,67 @@
+				</div>
+			}
+		>
+			{ phone.loaded ? (
+				<VerticalGapLayout size="large">
+					<PhoneNumberCard
+						phoneNumber={ phone }
+						initEditing={ initEditing }
+						onPhoneNumberChange={ onPhoneNumberChange }
+					/>
+					<div>TODO: add store address card</div>
+				</VerticalGapLayout>
+			) : (
+				<AppSpinner />
+			) }
```
Speaking of https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681156476 I think this loading state handling is redundant, and delivers slightly worse UX.

We can remove the redundancy from `ContactInformation` and let the `PhoneNumberCard` handle loading state itself.
```suggestion
			<VerticalGapLayout size="large">
				<PhoneNumberCard
					phoneNumber={ phone }
					initEditing={ initEditing }
					onPhoneNumberChange={ onPhoneNumberChange }
				/>
				<div>TODO: add store address card</div>
			</VerticalGapLayout>
```

tomalec replied:
After the change suggested above, we would have to change `initEditing` handling.
Personally, in `ContactInformation` I'd cater for `isSetupMC` only. Then let `PhoneNumberCard` reason about `phone.data.isValid` + some given property.
.

eason9487 replied:
Compare this way to https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681653834, it could only deliver the phone icon and "Phone number" title earlier about 0.5 ~ 1.5 seconds at the most. The rest elements are still needed to wait for data loaded. I'm not sure if that is worth it.

tomalec replied:
I'd say 0.5~1.5 s on a developer machine is a lot. Then imagine how long it could take on a low-end device with a slow network.

> I'm not sure if that is worth it.

Personally, I see it the opposite, compare:
- Deliver a scaffold instantly, with less code, and a single component that's responsible for the preloading state of its data
- Deliver a blank space (or card), with more code, and preloading logic and responsibility scattered around 2 components (3 in #921)

eason9487 replied:
Opened another PR #934 to change the handling of loading states.

tomalec reviewed (approved):
LGTM, delivers the needed feature. I think it's good to merge to unlock the following PRs, as we would touch on preloading changes later in #921. 

We can get back to the discussion from https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681156476 there or in a separate issue :scroll: hopefully soon, until we build too much on top of that, making it hard to change.

I'd also appreciate @jconroy take on evaluating npm dependencies and policy for sending PII to external libraries.
